### DISCUSSION TITLE
Duplicate operationId in OpenAPI schema

### DISCUSSION THREAD

**Discussion by MarkusH** [encode/django-rest-framework#8294] [Potential Issue] [answered]
After upgrading from 3.12.4 to 3.13.0 I noticed that some `operationId`s in the OpenAPI schema are duplicated. Using the `generateschema` command I also get warnings:

```
 django-admin generateschema --file openapi.json --format openapi-json
/tmp/django-rest-framework/rest_framework/schemas/openapi.py:49: UserWarning: You have a duplicated operationId in your OpenAPI schema: listWrappeds
        Route: /api/v1/a/, Method: get
        Route: /api/v1/b/, Method: get
        An operationId has to be unique across your schema. Your schema may not work in other tools.
  warnings.warn(
/tmp/django-rest-framework/rest_framework/schemas/openapi.py:49: UserWarning: You have a duplicated operationId in your OpenAPI schema: listWrappeds
        Route: /api/v1/b/, Method: get
        Route: /api/v1/c/, Method: get
        An operationId has to be unique across your schema. Your schema may not work in other tools.
  warnings.warn(
/tmp/django-rest-framework/rest_framework/schemas/openapi.py:49: UserWarning: You have a duplicated operationId in your OpenAPI schema: listWrappeds
        Route: /api/v1/c/, Method: get
        Route: /api/v1/d/, Method: get
        An operationId has to be unique across your schema. Your schema may not work in other tools.
  warnings.warn(
```

Digging deeper, this is cause by this change: https://github.com/encode/django-rest-framework/commit/9c97946531b85858fcee5df56240de6d29571da2

Looking at the implementation of `rest_framework.schema.openapi.AutoSchema.get_operation_id_base()`, this is somehow related to the `else:` branch where a trailing `APIView` string is stripped.

However, looking at the repr of the same view on 3.12.4 and 3.13.0, they look differently:

```python
# 3.12.4
In [1]: from myapp.views import my_view

In [2]: my_view.view_class
Out[2]: myapp.views.WrappedAPIView

In [3]: my_view.view_class.__name__
Out[3]: 'my_view'
```
```python
# 3.13.0
In [1]: from myapp.views import my_view

In [2]: my_view.view_class
Out[2]: rest_framework.decorators.api_view.<locals>.decorator.<locals>.WrappedAPIView

In [3]: my_view.view_class.__name__
Out[3]: 'WrappedAPIView'
```

**Comment by lovelydinosaur** [comment, marked as answer]
Okay - I think a possibly sensible path here is:

* Revert https://github.com/encode/django-rest-framework/pull/8291 (Which I assume is what's caused the behaviour change here.)
* Investigate in more detail *exactly* what issues we have with Django 4.0 that prompted #8291, and address those in a way that doesn't cause this breakage.

**Reply by MarkusH** [reply]
Thanks @tomchristie. Do you want me to open a MR to revert the commit? Should I turn the topic into a ticket?

**Reply by lovelydinosaur** [reply]
Not sure - having a look into it now

**Reply by lovelydinosaur** [reply]
Okay, just straight reverted it for now - 3.13.1.
That's got to be the best course of action to start with.

**Reply by MarkusH** [reply]
:heart: Thank you @tomchristie . I'll take a shot at the new version.

**Reply by MarkusH** [reply]
Confirm, 3.13.1 fixes the problem described above. Thanks!
